### PR TITLE
Add support for BIO callbacks in read, write, and free.

### PR DESCRIPTION
**Pull request by andrewhop**
### Issues:
Resolves CryptoAlg-953

### Description of changes: 
This adds support for legacy `callback`s and newer `callback_ex`s in `BIO_read`, `BIO_write`, and `BIO_free`. These callbacks provide the same API as OpenSSL 1.1.1 and 3.0.0 with the same semantics.

### Callouts
I'm not sure if reusing the logic in `call_bio_callback_with_processed` is helping or hurting the readability. I'm interested in feedback if it people prefer keeping that logic in `BIO_read` and `BIO_write`.

### Testing:
I created a simple mock callback to test this. I am looking for any better examples to make sure I implemented the behavior as expected. Right now the unit test is tightly coupled to the implementation and what my understanding of OpenSSL's code is doing. [bio.h](https://github.com/openssl/openssl/blob/master/include/openssl/bio.h.in), [bio_lib.c](https://github.com/openssl/openssl/blob/master/crypto/bio/bio_lib.c), and [docs](https://www.openssl.org/docs/man3.0/man3/BIO_set_callback_ex.html) for reference.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by bryce-shang — COMMENTED** [review]
> This adds support for legacy callbacks and newer callback_exs in BIO_read, BIO_write, and BIO_free. These callbacks provide the same API as OpenSSL 1.1.1 and 3.0.0 with the same semantics.

Nit: the original TT `P50557436` shows the package may only need/depend on 1.0.x. I may miss something.

File: include/openssl/bio.h [hunk 1]
```diff
@@ -865,6 +906,17 @@ struct bio_method_st {
...
   const BIO_METHOD *method;
 
+  // If set, |BIO_read|, |BIO_write|, and |BIO_free| execute |callback| or
+  // |callback_ex|, with a preference for the latter. Callbacks are only called
+  // with for the following events: |BIO_CB_READ|, |BIO_CB_READ|+|BIO_CB_RETURN|,
+  // BIO_CB_WRITE|, |BIO_CB_WRITE|+|BIO_CB_RETURN|, and |BIO_CB_FREE|.
+  long (*callback) (struct bio_st *bio, int oper, const char *argp, int argi,
+                   long argl, long bio_ret);
+  long (*callback_ex)(BIO *bio, int oper, const char *argp, size_t len,
+                      int argi, long argl, int bio_ret, size_t *processed);
```
[the remark this answers is not in the document]
Comment (andrewhop) [reply include/openssl/bio.h]: Removed the `bio_st` version all together.

File: crypto/bio/bio.c [hunk 2]
```diff
@@ -68,6 +68,72 @@
...
+  // Strip off any BIO_CB_RETURN flag
+  int bareoper = oper & ~BIO_CB_RETURN;
+
+  // We have an old style callback, so we will have to do nasty casts and
+  // check for overflows.
+  if (HAS_LEN_OPER(bareoper)) {
+    // In this case |len| is set, and should be used instead of |argi|
+    if (len > INT_MAX) {
+      return -1;
+    }
```
[the remark this answers is not in the document]
Comment (andrewhop) [reply crypto/bio/bio.c]: All of this got simplified and removed.

File: crypto/bio/bio_test.cc [hunk 3]
```diff
@@ -322,4 +322,117 @@ TEST_P(BIOPairTest, TestPair) {
   EXPECT_EQ(Bytes("12345"), Bytes(buf, 5));
 }
 
+#define CALL_BACK_FAILURE -1234567
```
[the remark this answers is not in the document]
Comment (andrewhop) [reply crypto/bio/bio_test.cc:325]: I wanted a unique value to differentiate from other failures :shrug:

File: crypto/bio/bio_test.cc [hunk 4]
```diff
@@ -322,4 +322,117 @@ TEST_P(BIOPairTest, TestPair) {
...
+  EXPECT_EQ(TEST_DATA_WRITTEN, BIO_write(bio1, "12345", TEST_DATA_WRITTEN));
+  ASSERT_EQ(TEST_DATA_WRITTEN, BIO_read(bio2, buf, sizeof(buf)));
+  EXPECT_EQ(Bytes("12345"), Bytes(buf, TEST_DATA_WRITTEN));
+
+  // Check that read or write was called first, then the combo with BIO_CB_RETURN
+  ASSERT_EQ(param_oper_ex[0], BIO_CB_READ);
+  ASSERT_EQ(param_oper_ex[1], BIO_CB_READ | BIO_CB_RETURN);
+
+  // argp is a pointer to a buffer for read/write operations. We don't care
+  // where the buf is, but it should be the same before and after the BIO calls
```
[the remark this answers is not in the document]
Comment (andrewhop) [reply crypto/bio/bio_test.cc:392]: This is checking that the same buffer is used in both calls, so the addresses match.

**Review by torben-hansen — APPROVED** [review]

**Review by bryce-shang — APPROVED** [review]

**Pull request merged**